packages/pglite-sync/src/index.ts

```diff
@@ -113,123 +156,183 @@ async function createPlugin(
-                      message.headers.operation === 'insert'
-                    ) {
-                      initialInserts.push(message as InsertChangeMessage)
-                    } else {
-                      foundNonInsert = true
-                      remainingMessages.push(message)
-                    }
-                  }
-                  if (initialInserts.length > 0) {
-                    // As `COPY FROM` doesn't trigger a NOTIFY, we pop
-                    // the last insert message and and add it to the be beginning
-                    // of the remaining messages to be applied after the `COPY FROM`
-                    remainingMessages.unshift(initialInserts.pop()!)
-                  }
-                  messageAggregator = remainingMessages
-
-                  // Do the `COPY FROM` with initial inserts
-                  if (initialInserts.length > 0) {
-                    applyMessagesToTableWithCopy({
-                      pg: tx,
-                      table: options.table,
-                      schema: options.schema,
-                      messages: initialInserts as InsertChangeMessage[],
-                      mapColumns: options.mapColumns,
-                      primaryKey: options.primaryKey,
-                      debug,
-                    })
-                    // We don't want to do a `COPY FROM` again after that
-                    doCopy = false
-                  }
-                }
-
-                for (const changeMessage of messageAggregator) {
-                  await applyMessageToTable({
-                    pg: tx,
-                    table: options.table,
-                    schema: options.schema,
-                    message: changeMessage,
-                    mapColumns: options.mapColumns,
-                    primaryKey: options.primaryKey,
-                    debug,
-                  })
-                }
 
+            if (options.commitGranularity === 'operation') {
+              // commit after each operation if granularity is set to operation
+              await throttledCommit()
+            } else if (typeof options.commitGranularity === 'number') {
+              // commit after every N messages if granularity is set to a number
+              if (messageAggregator.length >= options.commitGranularity) {
+                await throttledCommit()
+              }
+            }
+          } else if (isControlMessage(message)) {
+            switch (message.headers.control) {
+              case 'must-refetch':
+                // mark table as needing truncation before next batch commit
+                if (debug) console.log('refetching shape')
+                truncateNeeded = true
+                messageAggregator = []
+                break
+
+              case 'up-to-date':
+                // perform all accumulated changes and store stream state
+                await commit() // not throttled, we want this to happen ASAP
```

Comment:
this does not reset the throttle timer though - perhaps the throttled call should have a `reset` option or something?